spec/abci++/abci++_properties_001_draft.md

```diff
@@ -0,0 +1,675 @@
+    * Likewise, the Application must keep all responses to executing transactions until it can call `ResponseFinalizeBlock`.
+    * Application can change the transaction list via `ResponsePrepareProposal.tx`.
+      See [TransactionRecord](#transactionrecord) for further information on how to use it. Some notes:
+        * To remove a transaction from the proposed block the Application _marks_ the transaction as
+          invalid. It does not remove it from the list.
+        * Removing a transaction from the list means it is too early to propose that transaction,
+          so it will stay in the mempool.
+          The Application should be extra-careful with this feature to avoid leaks in the mempool.
+        * The `old_hash` field, besides helping with mempool maintenance, helps Tendermint handle
+          queries such as "what happened with this Tx?", by answering "it was modified into this one".
+        * The Application _can_ reorder the transactions in the list.
+    * The implementation of `PrepareProposal` can be non-deterministic.
+
+#### When does Tendermint call it?
+
+When a validator _p_ enters Tendermint consensus round _r_, height _h_, in which _p_ is the proposer:
+
+1. If _p_'s _validValue_
+    * is not `nil`
+      * _p_'s Tendermint uses the block stored in _validValue_ as block _v_
+    * is `nil`
+      * _p_'s Tendermint collects outstanding transactions from the mempool
+        * The transactions will be collected in order of priority
+        * Let $C$ the list of currently collected transactions
+        * The collection stops when any of the following conditions no longer holds
+          * the mempool is not empty
+          * the total size of transactions $\in C$ is less than `consensusParams.block.max_bytes`
+          * the sum of `GasWanted` field of transactions $\in C$ is less than `consensusParams.block.max_gas`
+      * _p_'s Tendermint creates a block header.
+      * _p_'s Tendermint uses the newly generated block as block _v_
+2. _p_'s Tendermint calls `RequestPrepareProposal` with block _v_.
+   The call is synchronous: Tendermint's execution will block until the Application returns from the call.
+3. The Application checks the block (header, transactions, commit info, evidences). Besides,
+   * in "same-block execution" mode, the Application can (and should) provide `ResponsePrepareProposal.data`,
+     `ResponsePrepareProposal.validator_updates`, or
+     `ResponsePrepareProposal.consensus_param_updates`.
+   * in "next-block execution" mode, _p_'s Tendermint will ignore the values for `ResponsePrepareProposal.data`,
+     `ResponsePrepareProposal.validator_updates`, and `ResponsePrepareProposal.consensus_param_updates`.
+   * in both modes, the Application can manipulate transactions
+      * leave transaction untouched
+      * add new transactions (not previously in the mempool)
+      * mark transactions is invalid (remove)
+      * remove transactions (effectively _delaying_ them)
```

Comment:
I see this now. You can mark it as `REMOVE` or remove it from the list. I think that the text here is a little ambiguous and that my preference is better captured in my comment above: We should make the `modifications` a list of just `modifications` and if a tx is absent from the list of modifications, then it will be proposed _as is_ by tendermint.

Reply:
See my replay above about reordering.